### PR TITLE
[Deepspeed] deepspeed auto grad accum

### PR DESCRIPTION
**Pull request by kashif**
# What does this PR do?

add automatic gradient accumulation boundary handling for deepspeed and non-deepspeed setting
and get the gradient norms for deepspeed

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/accelerate/blob/main/CONTRIBUTING.md#submitting-a-pull-request-pr),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/accelerate/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/accelerate/tree/main/docs#writing-documentation---specification).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: docs/source/usage_guides/deepspeed.md [hunk 1]
```diff
@@ -631,6 +631,109 @@ ds_config: {'bf16': {'enabled': False}, 'zero_optimization': {'stage': 3, 'stage
...
+**After (unified approach):**
+```python
+# This approach works identically for both DeepSpeed and standard training!
+with accelerator.accumulate(model):
+    outputs = model(inputs)
+    loss = loss_fn(outputs, targets)
+    
+    # Handles DeepSpeed boundaries automatically!
+    accelerator.backward(loss)
+
```
Comment (SunMarc) [inline docs/source/usage_guides/deepspeed.md]: indeed this is a bit out of place, when ppl use accelerate, they indeed expect things to work directly by using accelerator.backward so on so forth. Not sure we need to remind them again.

File: tests/deepspeed/test_deepspeed_gradient_accumulation.py [hunk 2]
```diff
@@ -0,0 +1,288 @@
...
+            MASTER_ADDR="localhost",
+            MASTER_PORT="10999",
+            RANK="0",
+            LOCAL_RANK="0",
+            WORLD_SIZE="1",
+        )
+
+    def get_config_dict(self, stage):
+        # As some tests modify the dict, always make a copy
+        return deepcopy(self.ds_config_dict[stage])
```
Comment (SunMarc) [inline tests/deepspeed/test_deepspeed_gradient_accumulation.py]: not used

File: src/accelerate/utils/deepspeed.py [hunk 3]
```diff
@@ -261,22 +261,38 @@ class DeepSpeedEngineWrapper:
...
-        # Deepspeed's `engine.step` performs the following operations:
-        # - gradient accumulation check
-        # - gradient clipping
-        # - optimizer step
-        # - zero grad
-        # - checking overflow
-        # - lr_scheduler step (only if engine.lr_scheduler is not None)
-        self.engine.step()
+        # Only perform step and related operations at gradient accumulation boundaries
+        if sync_gradients:
```
Comment (SunMarc) [inline src/accelerate/utils/deepspeed.py:273]: just to be sure, this is optional no ? deepspeed should be able to figure out to step or not. We can keep it ofc

File: src/accelerate/utils/deepspeed.py [hunk 4]
```diff
@@ -261,22 +261,38 @@ class DeepSpeedEngineWrapper:
     def __init__(self, engine):
         self.engine = engine
 
-    def backward(self, loss, **kwargs):
+    def backward(self, loss, sync_gradients=True, **kwargs):
+        # Set gradient accumulation boundary based on Accelerate's sync_gradients state
+        # This tells DeepSpeed whether this is the final micro-batch before gradient sync
+        self.engine.set_gradient_accumulation_boundary(is_boundary=sync_gradients)
+
```
Comment (SunMarc) [inline src/accelerate/utils/deepspeed.py:268]: just to be sure when was this introduced ? I just want to make sure that eveyrhting is backward compatible with older version of deepspeed

Comment (kashif) [reply src/accelerate/utils/deepspeed.py:268]: i also checked and this was introduced 4+ years ago

Comment (SunMarc) [reply src/accelerate/utils/deepspeed.py:268]: SG !

Comment (kashif) [reply src/accelerate/utils/deepspeed.py:268]: https://github.com/deepspeedai/DeepSpeed/pull/1588

**Review by SunMarc — APPROVED** [review]
Thanks a lot !

**Comment by SunMarc** [comment]
@bot /style

**Comment by skandermoalla** [comment]
@kashif what was the issue before? Is this a breaking change?

**Comment by kashif** [comment]
no i believe there is no issue... just previously folks had to handle deepspeed stuff kind of seperately and this conslidates the accerelate api to handle deepspeed backend so that previously one had to do:

```
def backward_step(loss, is_final_micro_batch=False):
    if using_deepspeed:
        model.set_gradient_accumulation_boundary(is_final_micro_batch)
        model.backward(loss)
        if is_final_micro_batch:
            model.step()
    else:
        accelerator.backward(loss)
        if accelerator.sync_gradients:
            optimizer.step()
            optimizer.zero_grad()
```

and one can still do that of course... but now its handled by accerelate so one can do:

```
with accelerator.accumulate(model):
    outputs = model(inputs)
    loss = loss_fn(outputs, targets)
    
    accelerator.backward(loss)
    
if accelerator.sync_gradients:
    accelerator.clip_grad_norm_(model.parameters(), max_norm=1.0)
    optimizer.step()
    optimizer.zero_grad()
```
hope that helps

**Comment by skandermoalla** [comment]
Awesome, thanks a lot! Do you know if Trainers in TRL are handling DeepSpeed separately in this way as well?

**Comment by kashif** [comment]
@skandermoalla yes the trainers are using the Transformers Trainers, which is doing the right thing, albeit using the API that handles the deepspeed seperately

**Comment by DDDPG** [comment]
Hi I want to know that if I use the older version of acc(v1.0.1), and applied with the deepspeed framework, how can I get the grad norm from the `clip_grad_norm_`? I saw that the gradient-related calculation hosted by deepspeed itself, acc leave the return value as none, how could I get it? Thanks a lot!
